### PR TITLE
Fix check that mappings can only have storage location

### LINKED ISSUE #5929: Mapping as function parameter with memory storage compiles
The following code compiles even though the mapping has the `memory` data location.

```
contract c {
        mapping(uint => uint) map;
    function f3(mapping(uint => uint) memory _mymapping) view internal {}
    function f2() public view {
        f3(map);
    }
}
```

### PR DESCRIPTION
**Pull request by Marenz**
Fixes #5929

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -334,10 +334,20 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
...
+			{
+				if (!isLibraryFunction && _function.isPublic())
+					m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\" and thus can for parameters or return variables only be used in internal or library functions.");
+				else
+					m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\"." );
+
+			}
+			else
+			{
+				// This situation is expected to be caught earlier by the data location checks
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Not sure we need the comment.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -334,10 +334,20 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
...
-				m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\"." );
-			else if (!isLibraryFunction && _function.isPublic())
-				m_errorReporter.typeError(var.location(), "Mapping types for parameters or return variables can only be used in internal or library functions.");
+			if (var.referenceLocation() != VariableDeclaration::Location::Storage)
+			{
+				if (!isLibraryFunction && _function.isPublic())
+					m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\" and thus can for parameters or return variables only be used in internal or library functions.");
+				else
+					m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\"." );
+
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Please remove the empty line.

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -334,10 +334,20 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
...
 		if (type(var)->category() == Type::Category::Mapping)
 		{
-			if (!type(var)->dataStoredIn(DataLocation::Storage))
-				m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\"." );
-			else if (!isLibraryFunction && _function.isPublic())
-				m_errorReporter.typeError(var.location(), "Mapping types for parameters or return variables can only be used in internal or library functions.");
+			if (var.referenceLocation() != VariableDeclaration::Location::Storage)
+			{
+				if (!isLibraryFunction && _function.isPublic())
+					m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\" and thus can for parameters or return variables only be used in internal or library functions.");
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
					m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\" and thus only be parameters or return variables for internal or library functions.");
```

**Review by chriseth — CHANGES_REQUESTED** [review]

**Comment by Marenz** [comment]
Updated according to comments

**Comment by chriseth** [comment]
Please add a changelog entry!

**Comment by Marenz** [comment]
Added changelog

**Comment by Marenz** [comment]
Fixed failing CI

**Review by chriseth — APPROVED** [review]

File: Changelog.md [hunk 4]
```diff
@@ -3,6 +3,7 @@
 Bugfixes:
  * Code generator: Defensively pad allocation of creationCode and runtimeCode to multiples of 32 bytes.
  * Parser: Disallow empty import statements.
+ * Type Checker: Dissallow mappings with data locations other than 'storage'
```
Comment (axic) [inline Changelog.md:6]: Typo: Dissallow :)

And we usually use double-backticks for keywords: 
```
``storage``
```

**Pull request merged**